### PR TITLE
#326 - Use scenario-specified barbarian units

### LINKED ISSUE #326: Use scenario-specified default barbarian units
Some scenarios use custom barbarian untis, e.g. 06 Age of Discovery and 07 Sengoku.  Our code fails because it always looks for Warrior and Galley, and these scenarios don't have those units, but others.

We should use those units instead.

### PR DESCRIPTION
**Pull request by QuintillusCFC**
This allows scenarios that don't use Warrior/Horseman/Galley for barbarians to work in that regard.

Notably, 06 MP Age of Discovery now loads with Archer barbarians.

This work spawned #329 and #330 to make further improvements to this area.

---

Note: With this change, the whole list of unit prototypes is imported (rather than a whitelisted subset), and the limiter of what can be built is moved to CanBuildCity in City.cs.  Eventually this will disappear, but this change allows the barbarians to build units that aren't on that whitelist.

It should also be noted that 06 MP Age of Discovery is not really playable yet, since you can only build Settlers and Workers and that rapidly leads to crashes due to running out of citizens.  But it can load its barbarians properly, and a custom scenario with modified barbarians should work too.

### CODE REVIEW

File: C7/Game.cs [hunk 1]
```diff
@@ -108,7 +108,13 @@ public override void _Ready()
...
 			errorOnLoad = true;
 			PopupOverlay popupOverlay = GetNode<PopupOverlay>(PopupOverlay.NodePath);
-			popupOverlay.ShowPopup(new ErrorMessage(ex.Message), PopupOverlay.PopupCategory.Advisor);
+			string message = ex.Message;
+			string[] stack = ex.StackTrace.Split("\r\n");	//for some reason it is returned with \r\n in the string as one line.  let's make it readable!
+			foreach (string line in stack) {
+				message = message + "\r\n" + line;
+			}
+
+			popupOverlay.ShowPopup(new ErrorMessage(message), PopupOverlay.PopupCategory.Advisor);
```
Comment (QuintillusCFC) [inline C7/Game.cs:117]: This gives a more useful error message if a file fails to load for some reason.  Yeah, it's in the logs too (which is nice!), but having a hint in-your-face is nice too.  Eventually we probably want to roll this back but when it's mostly developers playing, I think it makes sense.

File: C7GameData/ImportCiv3.cs [hunk 2]
```diff
@@ -217,50 +223,48 @@ public static C7SaveFormat ImportBiq(string biqPath, string defaultBiqPath)
...
+					prototype.actions.Add("fortify");
+				}
+				if (prto.Disband) {
+					prototype.actions.Add("disband");
+				}
+				if (prto.GoTo) {
+					prototype.actions.Add("goTo");
+				}
+				//Temporary check until #329/#330 are finished
+				if (!c7SaveFormat.GameData.unitPrototypes.ContainsKey(prototype.name)) {
```
Comment (QuintillusCFC) [inline C7GameData/ImportCiv3.cs:270]: This temporary check is the only change to this section, the rest is indentation.  This change allows scenarios with strategy maps (which is nearly all of them) to be loaded now that all unit prototypes are imported.

**Comment by WildWeazel** [comment]
bumping to resolve conflicts

**Comment by QuintillusCFC** [comment]
Fixed the conflicts.  Still looks sensible to me.  I'd kind of forgotten that this was still here and not merged.

**Comment by QuintillusCFC** [comment]
I'm going to go ahead and merge this since I self-reviewed it 3 days ago and there hasn't been another review since 7/23.  Also because I was looking at #140 and that will be easier once these changes are merged due to more scenarios being openable.

**Pull request merged**